akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala

```diff
@@ -246,5 +137,185 @@ private[akka] object AsyncDnsResolver {
+  private class RequestIdInjector extends Actor {
+    private var requestId: Short = 0
+
+    private def nextId(): Short = {
+      requestId = (requestId + 1).toShort
+      requestId
+    }
+
+    override def receive: Receive = {
+      case DnsQuestionPreInjection(resolver, inject, timeout) =>
+        implicit val tmout = timeout
+        val question = inject(nextId())
+        val forwardAnswerTo = sender()
+        val result = (resolver ? question).mapTo[Answer]
+
+        result.onComplete {
+          case Success(answer) => forwardAnswerTo ! answer
+          case Failure(_)      => resolver ! DropRequest(question.id)
+        }(ExecutionContexts.parasitic)
+    }
+  }
+
+  private object DnsResolutionActor {
+    def props(
+        settings: DnsSettings,
+        requestIdInjector: ActorRef,
+        name: String,
+        mode: RequestType,
+        answerTo: ActorRef,
+        cacheActor: ActorRef): Props =
+      Props(new DnsResolutionActor(settings, requestIdInjector, name, mode, answerTo, cacheActor))
+
+    final case class Resolve(withResolvers: List[ActorRef])
+  }
+
+  private class DnsResolutionActor(
+      settings: DnsSettings,
+      requestIdInjector: ActorRef,
+      name: String,
+      mode: RequestType,
+      answerTo: ActorRef,
+      cacheActor: ActorRef)
+      extends Actor
+      with ActorLogging {
+    import DnsResolutionActor._
+
+    private implicit val timeout: Timeout = Timeout(settings.ResolveTimeout)
+
+    private def failToResolve(): Unit =
+      answerWithFailure(AsyncDnsResolver.failToResolve(name, settings.NameServers))
+
+    private def answerWithFailure(ex: Throwable): Unit = answer(Status.Failure(ex))
+
+    private def answer(withMsg: Any): Unit = {
+      answerTo ! withMsg
+
+      context.stop(self)
+    }
+
+    // the fully-qualified domain names (FQDNs), in the order we want to attempt for any given resolver
+    private val namesToResolve = {
+      val nameWithSearch = settings.SearchDomains.map(sd => s"${name}.${sd}")
+
+      // ndots is a heuristic used to attempt to work out if `name` is an FQDN or a name relative to a search
+      // name.  The goal is to not incur the cost of a lookup that's unlikely to resolve because we need to
+      // relativize it.  If the name being searched contains less than ndots dots, then we look it up last,
+      // rather than first.
+      if (name.count(_ == '.') >= settings.NDots) {
+        // assume fully-qualified, so try `name` first
+        (name :: nameWithSearch).map(Helpers.toRootLowerCase)
+      } else {
+        // assume relative, so try `name` last
+        (nameWithSearch :+ name).map(Helpers.toRootLowerCase)
+      }
+    }
+
+    override def receive: Receive = {
+      case Resolve(withResolvers) =>
+        if (withResolvers.nonEmpty) startResolution(namesToResolve, withResolvers.head, withResolvers.tail)
+        else failToResolve() // (vacuously) exhausted the resolvers
```

Comment:
Let's skip the initial Resolve message as well and trigger resolution from constructor parameters on startup, no need to go via mailbox for that.

Reply:
I decided against that out of not wanting to hang on to the full list of resolvers for the lifetime of the actor (unlike the names to search, we never go back to start from the top).  Could also do it with a `var`, I guess.

Reply:
The list is of resolvers is anyway kept in memory (for the lifetime of the ActorSystem) in the parent AsyncDnsResolver.